### PR TITLE
Remove cudf._lib.orc in favor of inlining pylibcudf

### PR DESCRIPTION
**Pull request by mroeschke**
## Description
Contributes to https://github.com/rapidsai/cudf/issues/17317

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/utils/ioutils.py [hunk 1]
```diff
@@ -1487,6 +1498,153 @@
...
+            )
+        elif isinstance(
+            col.dtype,
+            (cudf.ListDtype, cudf.StructDtype, cudf.core.dtypes.DecimalDtype),
+        ):
+            types.append(col.dtype.to_arrow())
+        else:
+            # A boolean element takes 8 bits in cudf and 1 bit in
+            # pyarrow. To make sure the cudf format is interperable
+            # in arrow, we use `int8` type when converting from a
```
Comment (wence-) [inline python/cudf/cudf/utils/ioutils.py:1550]:
```suggestion
            # pyarrow. To make sure the cudf format is interoperable
            # with arrow, we use `int8` type when converting from a
```

File: python/cudf/cudf/utils/ioutils.py [hunk 2]
```diff
@@ -9,31 +9,42 @@
...
 try:
     import fsspec.parquet as fsspec_parquet
-
 except ImportError:
     fsspec_parquet = None
 
+try:
+    import ujson as json  # type: ignore[import-untyped]
+except ImportError:
+    import json
```
Comment (wence-) [inline python/cudf/cudf/utils/ioutils.py:35]: *question*: Why this? I think we're only using this to load the (tiny) metadata json dict, so it seems perhaps unnecessary?

File: python/pylibcudf/pylibcudf/io/types.pyx [hunk 3]
```diff
@@ -288,12 +288,14 @@ cdef class TableInputMetadata:
...
     def __init__(self, Table table):
         self.c_obj = table_input_metadata(table.view())
-        self.column_metadata = [
+
+    @property
+    def column_metadata(self):
+        return [
             ColumnInMetadata.from_libcudf(&self.c_obj.column_metadata[i], self)
             for i in range(self.c_obj.column_metadata.size())
         ]
```
Comment (wence-) [inline python/pylibcudf/pylibcudf/io/types.pyx:298]: *question*: What is this change needed for?

**Review by wence- — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/utils/ioutils.py (hunk 2 above)
Comment (mroeschke) [reply python/cudf/cudf/utils/ioutils.py:35]: I'm fine with just using `json`; I was just porting over the equivalent behavior

File: python/pylibcudf/pylibcudf/io/types.pyx (hunk 3 above)
Comment (mroeschke) [reply python/pylibcudf/pylibcudf/io/types.pyx:298]: Before this was an (unused) `cdef` attribute, but cudf Python ORC handling modifies the `column_metadata` inplace, so I needed to expose this attribute in Python space.

Comment (wence-) [reply python/pylibcudf/pylibcudf/io/types.pyx:298]: Can you add it as a property in types.pyi if it is not already done?

**Review by wence- — APPROVED** [review]

File: python/pylibcudf/pylibcudf/io/types.pyx (hunk 3 above)
Comment (mroeschke) [reply python/pylibcudf/pylibcudf/io/types.pyx:298]: Sure. Added in [5e27f2a](https://github.com/rapidsai/cudf/pull/17466/commits/5e27f2ae07574b02a2006baafbe79c455e9a26f0)
